### PR TITLE
feat(functions): add DATE_FROM_PARTS, TIMESTAMP_FROM_PARTS and TIMESTAMP_TZ_FROM_PARTS

### LINKED ISSUE #19435: Feature: Support Date & Timestamp Construction functions
**Summary**

Creates a `date`, `timestamp`, `timestamp_tz` from individual numeric components.

```
DATE_FROM_PARTS( <year>, <month>, <day> )
TIMESTAMP_FROM_PARTS( <year>, <month>, <day>, <hour>, <minute>, <second> [, <nanosecond> ] [, <time_zone> ] )
TIMESTAMP_TZ_FROM_PARTS( <year>, <month>, <day>, <hour>, <minute>, <second> [, <nanosecond>] [, <time_zone>] )
```

- [ ] DATE_FROM_PARTS
- [ ] TIMESTAMP_FROM_PARTS
- [ ] TIMESTAMP_TZ_FROM_PARTS

reference
https://docs.snowflake.com/en/sql-reference/functions/date_from_parts
https://docs.snowflake.com/en/sql-reference/functions/timestamp_from_parts

### PR DESCRIPTION
**Pull request by tharunn0**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

This PR introduces support for the following datetime construction functions:

- DATE_FROM_PARTS(year, month, day)
- TIMESTAMP_FROM_PARTS(year, month, day, hour, minute, second [, nanosecond] [, time_zone])
- TIMESTAMP_TZ_FROM_PARTS(...)

These functions construct date and timestamp values from individual numeric components.

Integration tests and sqllogictests were added to verify correctness.

Fixes #19435

## Type of change

- [x] New Feature (non-breaking change which adds functionality)

## Tests

- [x] Unit tests
- [x] Integration tests
- [x] SQLLogicTests

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/19508)

### CODE REVIEW

**Comment by b41sh** [comment]
Hi @tharunn0 
Thanks for your contribution, please run `cargo fmt` and `cargo clippy` to fix the CI

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 1]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+    match len {
+        Some(_) => Value::Column(Column::Timestamp(builder.into())),
+        _ => Value::Scalar(Scalar::Timestamp(builder[0])),
+    }
+}
+
+fn register_timestamp_from_parts(registry: &mut FunctionRegistry) {
+    registry.register_aliases("timestamp_from_parts", &[
+        "timestampfromparts",
+        "timestamp_ntz_from_parts",
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: "timestamp_ntz_from_parts",
        "timestampntzfromparts",
        "timestamp_ltz_from_parts",
        "timestampltzfromparts",

Why support these alias? Databend only support timestamp type and timestamp_tz. So you should delete them.

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 2]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+        .to_zoned(dt)
+        .map_err(|e| format!("Cannot construct timestamp: {e}"))?;
+    Ok(zoned.timestamp().as_microsecond())
+}
+
+fn register_date_from_parts(registry: &mut FunctionRegistry) {
+    registry.register_aliases("date_from_parts", &["datefromparts"]);
+
+    registry
+        .register_passthrough_nullable_3_arg::<Int32Type, Int32Type, Int32Type, DateType, _, _>(
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: If you use Int32Type as arg, will generate these bug:

```sql
:) select date_from_parts(2004::int64, 1, 1);
error: APIError: QueryFailed: [1065]error: 
  --> SQL:1:8
  |
1 | select date_from_parts(2004::int64, 1, 1)
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no function matches signature `date_from_parts(Int64, UInt8, UInt8)`, you might need to add explicit type casts.

candidate functions:
  date_from_parts(Int32, Int32, Int32) :: Date                      : unable to unify `Int64` with `Int32`
  date_from_parts(Int32 NULL, Int32 NULL, Int32 NULL) :: Date NULL  : unable to unify `Int64` with `Int32`
```

Please replace to Int64Type. Databend has implicit conversion capabilities for the int64 type.

e.g.

```rust
registry.register_passthrough_nullable_2_arg::<TimestampType, Int64Type, DateType, _, _>(
        "to_start_of_week",
        |_, _, _| FunctionDomain::Full,
        vectorize_2_arg::<TimestampType, Int64Type, DateType>(|val, mode, ctx| {
            if mode == 0 {
                DateRounder::eval_timestamp::<ToLastSunday>(val, &ctx.func_ctx.tz)
            } else {
                DateRounder::eval_timestamp::<ToLastMonday>(val, &ctx.func_ctx.tz)
            }
        }),
    );
```

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 3]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+        );
+}
+
+fn timestamp_from_parts_fn(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
+    let len = args.iter().find_map(|arg| match arg {
+        Value::Column(col) => Some(col.len()),
+        _ => None,
+    });
+
+    let year_arg = args[0].try_downcast::<Int32Type>().unwrap();
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: Ditto. Int32Type -> Int64Type

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 4]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+    let result = base
+        .checked_add_signed(chrono::Duration::days(day as i64 - 1))
+        .ok_or_else(|| format!("Date out of range: year={year}, month={month}, day={day}"))?;
+
+    Ok(result)
+}
+
+/// Normalize timestamp components that may be outside normal ranges.
+/// Similar to normalize_date_parts but also handles hour/minute/second/nanosecond overflow.
+fn normalize_timestamp_micros(
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: refactor to src/query/expression/src/utils/date_helper.rs

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 5]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
     };
     replace_time_format(&processed_format).to_string()
 }
+
+/// Normalize month/day values that may be outside normal ranges.
+/// Snowflake allows e.g. month=0 (meaning Dec of previous year),
+/// month=13 (meaning Jan of next year), day=100 (100th day from month start), etc.
+fn normalize_date_parts(year: i32, month: i32, day: i32) -> Result<NaiveDate, String> {
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: refactor to src/query/expression/src/utils/date_helper.rs. And it looks has same logic with normalize_timestamp_micros

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 6]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+        final_hour,
+        final_minute,
+        final_second,
+        norm_micros,
+    ) {
+        return Ok(micros);
+    }
+
+    // Slow path via jiff
+    let dt = jiff::civil::date(jiff_year, jiff_month, jiff_day).at(
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: use jiff::civil::date

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 7]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+    let norm_year = (total_months.div_euclid(12)) as i32;
+    let norm_month = (total_months.rem_euclid(12) + 1) as u32; // 1..=12
+
+    // Create date at day=1 of the normalized year/month, then add (day-1) days
+    let base = NaiveDate::from_ymd_opt(norm_year, norm_month, 1)
+        .ok_or_else(|| format!("Invalid date: year={year}, month={month}, day={day}"))?;
+
+    // day=1 means the first day, day=0 means one day before, day=-1 means two days before, etc.
+    let result = base
+        .checked_add_signed(chrono::Duration::days(day as i64 - 1))
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: Why use jiff in normalize_timestamp_micros and use chrono in normalize_date_parts?

And please use jiff crate api.

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 8]
```diff
@@ -2733,3 +2749,361 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+            FunctionDomain::MayThrow,
+            timestamp_from_parts_fn,
+            None,
+            has_null,
+        )))
+    }));
+    registry.register_function_factory("timestamp_from_parts", factory);
+}
+
+fn timestamp_tz_from_parts_fn(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: timestamp_from_parts_fn and timestamp_tz_from_parts_fn have a lot of repetitive logic

File: src/query/functions/tests/it/scalars/testdata/datetime.txt [hunk 9]
```diff
@@ -3873,3 +3873,122 @@ error:
...
+optimized expr : 1365163200000000
+output type    : Timestamp
+output domain  : {1365163200000000..=1365163200000000}
+output         : '2013-04-05 12:00:00.000000'
+
+
+error: 
+  --> SQL:1:1
+  |
+1 | timestamp_from_parts(2013, 4, 5, 12, 0, 0, 987654321)
```
Comment (TCeason) [inline src/query/functions/tests/it/scalars/testdata/datetime.txt]: Because int32type as arg type so it's not right.

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 10]
```diff
@@ -2733,3 +2749,360 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+    // Split into extra days and time-of-day
+    let extra_days = total_seconds.div_euclid(86400);
+    let time_of_day = total_seconds.rem_euclid(86400);
+
+    let final_hour = (time_of_day / 3600) as u8;
+    let final_minute = ((time_of_day % 3600) / 60) as u8;
+    let final_second = (time_of_day % 60) as u8;
+
+    // Adjust day by the overflow days
+    let final_day = day + extra_days as i32;
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: i32::try_from(extra_days).map_err(...) 

safe cast is better

File: tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test [hunk 11]
```diff
@@ -1812,3 +1812,118 @@ query ??
...
+2013-04-05 11:00:00.000000
+
+## TIMESTAMP_TZ_FROM_PARTS
+
+statement ok
+set timezone = 'UTC'
+
+# Basic timestamp_tz construction with default (UTC) timezone
+query T
+select typeof(timestamp_tz_from_parts(2013, 4, 5, 12, 0, 0));
```
Comment (TCeason) [inline tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test]: `timestamp_tz_from_parts` has insufficient coverage in the sqllogictest. 
The test only verified that `typeof()` returns `TIMESTAMP_TZ`, but did not validate the actual values. Missing:
- Tests with timezone parameters (7 parameters string, 8 parameters)
- 8-parameter test with nanosecond + timezone - verification of value correctness in different time zones
- Boundary value/overflow tests

And timestamp_from_parts also needs to add these test.

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 12]
```diff
@@ -2733,3 +2749,360 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+        let signature = FunctionSignature {
+            name: "timestamp_tz_from_parts".to_string(),
+            args_type,
+            return_type: DataType::TimestampTz,
+        };
+
+        Some(Arc::new(Function::with_passthrough_nullable(
+            signature,
+            FunctionDomain::MayThrow,
+            timestamp_tz_from_parts_fn,
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: The eval function re-discovers the argument layout at runtime via try_downcast, but the factory already knows it at registration time. If these two pieces of logic ever drift apart, you get a silent unwrap() panic that only triggers on specific argument combinations.                                                   
                                                                                                                                  
  Since the factory already determines whether arg[6] is nanosecond or timezone, just capture that decision and pass it through:
```rust

  let (sig_args, has_nano, has_tz) = match args_type.len() {
      6 => (vec![int32_type; 6], false, false),
      7 if args_type[6].remove_nullable() == DataType::String => {
          let mut v = vec![int32_type; 6];
          v.push(DataType::String);
          (v, false, true)
      }
      7 => (vec![int32_type; 7], true, false),
      8 => {
          let mut v = vec![int32_type; 7];
          v.push(DataType::String);
          (v, true, true)
      }
      _ => return None,
  };

  // closure captures has_nano/has_tz — eval doesn't guess
  move |args, ctx| timestamp_tz_from_parts_impl(args, ctx, has_nano, has_tz)
```
  Then in the eval function, index directly by known layout instead of trial-and-error downcasting.

Comment (TCeason) [reply src/query/functions/src/scalars/timestamp/src/datetime.rs]: Same pattern applies to `timestamp_from_parts` — though lower risk since all args are Int32, keeping the same approach avoids inconsistency.

**Review by TCeason — CHANGES_REQUESTED** [review]

File: tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test (hunk 11 above)
Comment (TCeason) [reply tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test]: After build databend-query with your pr, can run this test:

```
./target/debug/databend-sqllogictests   --run_file  02_0012_function_datetimes.test
```

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 13]
```diff
@@ -2733,3 +2749,264 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+            func,
+            None,
+            has_null,
+        )))
+    }));
+
+    registry.register_function_factory("timestamp_from_parts", factory);
+}
+
+fn timestamp_from_parts_6(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]: why extract timestamp_from_parts_7 and timestamp_from_parts_6 function? only one call.

File: tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test [hunk 14]
```diff
@@ -1812,3 +1812,84 @@ query ??
...
+
+# Negative seconds overflow with timezone
+query ?
+select timestamp_tz_from_parts(2013, 4, 5, 12, 0, -3600, 'UTC');
+----
+2013-04-05 11:00:00.000000 +0000
+
+# Month overflow with timezone
+query ?
+select timestamp_tz_from_parts(2013, 13, 1, 0, 0, 0, 'UTC');
```
Comment (TCeason) [inline tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test:1888]: I think you can use llm add more test case. For example you can add some test about dst with Asia/Shanghai and America/Los_Angeles timezone.

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 15]
```diff
@@ -2733,3 +2749,264 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+
+fn timestamp_tz_from_parts_6(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
+    timestamp_tz_from_parts_fn(args, ctx, false, false)
+}
+
+fn timestamp_tz_from_parts_7_nano(
+    args: &[Value<AnyType>],
+    ctx: &mut EvalContext,
+) -> Value<AnyType> {
+    timestamp_tz_from_parts_fn(args, ctx, true, false)
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs]:
```rust
fn timestamp_tz_from_parts_6(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
    timestamp_tz_from_parts_fn(args, ctx, false, false)
}

fn timestamp_tz_from_parts_7_nano(
    args: &[Value<AnyType>],
    ctx: &mut EvalContext,
) -> Value<AnyType> {
    timestamp_tz_from_parts_fn(args, ctx, true, false)
}

fn timestamp_tz_from_parts_7_tz(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
    timestamp_tz_from_parts_fn(args, ctx, false, true)
}

fn timestamp_tz_from_parts_8(args: &[Value<AnyType>], ctx: &mut EvalContext) -> Value<AnyType> {
    timestamp_tz_from_parts_fn(args, ctx, true, true)
}
```

I think can use timestamp_tz_from_parts_fn directly.

And comment them.

e.g.

```rust
8 => {
                let mut v = vec![int64_type; 7];
                v.push(DataType::String);
                // add some comment when arg len equal with 8
                (v, timestamp_tz_from_parts_fn(args, ctx, true, true) as _)
            }
```

**Review by TCeason — CHANGES_REQUESTED** [review]

File: src/query/expression/src/utils/date_helper.rs [hunk 16]
```diff
@@ -2066,3 +2067,109 @@ pub fn pg_format_to_strftime(pg_format_string: &str) -> String {
...
 }
+
+/// Normalize month/day values that may be outside normal ranges.
+/// Supports values like month=0 (meaning Dec of previous year),
+/// month=13 (meaning Jan of next year), and day=100 (100th day from month start).
+pub fn normalize_date_parts(year: i64, month: i64, day: i64) -> std::result::Result<Date, String> {
+    // 1. Safe month normalization
+    let total_months = year
+        .checked_mul(12)
+        .and_then(|y| y.checked_add(month - 1))
```
Comment (TCeason) [inline src/query/expression/src/utils/date_helper.rs:2078]: normalize_date_parts still does unchecked arithmetic on user inputs via month - 1 (and later day - 1) before the surrounding
    checked_* calls run. Inputs near the INT64 bounds, such as date_from_parts(1, -9223372036854775808, 1), will therefore panic
    in the repo's dev/test profiles and wrap to a bogus value in release builds instead of returning a normal query error; this
    also leaks into timestamp_from_parts/timestamp_tz_from_parts because they reuse the same helper.

File: src/query/expression/src/utils/date_helper.rs [hunk 17]
```diff
@@ -2066,3 +2067,109 @@ pub fn pg_format_to_strftime(pg_format_string: &str) -> String {
...
+    let extra_secs_from_nanos = nanosecond.div_euclid(1_000_000_000);
+    let norm_nanos = nanosecond.rem_euclid(1_000_000_000) as i32;
+    let norm_micros = (norm_nanos / 1_000) as u32;
+
+    // Combine time components into seconds
+    let total_seconds = hour
+        .checked_mul(3600)
+        .and_then(|h| h.checked_add(minute * 60))
+        .and_then(|m| m.checked_add(second))
+        .and_then(|s| s.checked_add(extra_secs_from_nanos))
```
Comment (TCeason) [inline src/query/expression/src/utils/date_helper.rs:2125]: The overflow handling in normalize_timestamp_micros is incomplete because minute * 60 is evaluated before it enters the
    checked_add chain. Large INT64 minute values will panic in checked builds and wrap in release builds, so timestamp_from_parts/
    timestamp_tz_from_parts can return an incorrect timestamp or crash instead of surfacing BadArguments for out-of-range
    components.

File: src/query/expression/src/utils/date_helper.rs [hunk 18]
```diff
@@ -2066,3 +2067,109 @@ pub fn pg_format_to_strftime(pg_format_string: &str) -> String {
 
     result
 }
+
+/// Normalize month/day values that may be outside normal ranges.
+/// Supports values like month=0 (meaning Dec of previous year),
+/// month=13 (meaning Jan of next year), and day=100 (100th day from month start).
+pub fn normalize_date_parts(year: i64, month: i64, day: i64) -> std::result::Result<Date, String> {
```
Comment (TCeason) [inline src/query/expression/src/utils/date_helper.rs:2074]: Why not return databend_common_exception::Result? Can return Errorcode::BadArgument

File: src/query/expression/src/utils/date_helper.rs [hunk 19]
```diff
@@ -2066,3 +2067,109 @@ pub fn pg_format_to_strftime(pg_format_string: &str) -> String {
...
+    let base = Date::new(norm_year, norm_month, 1)
+        .map_err(|_| format!("Invalid date: year={year}, month={month}, day={day}"))?;
+
+    // 2. Safe day normalization (Fixing the second bot flag)
+    let hours_to_add = (day - 1)
+        .checked_mul(24)
+        .ok_or_else(|| format!("Day value out of bounds: {day}"))?;
+
+    let result = base
+        .checked_add(SignedDuration::from_hours(hours_to_add))
```
Comment (TCeason) [inline src/query/expression/src/utils/date_helper.rs:2097]: Why not use checked_add(SignedDuration::from_days(day - 1))? And please use checked day-1

File: src/query/functions/src/scalars/timestamp/src/datetime.rs [hunk 20]
```diff
@@ -2733,3 +2749,261 @@ fn prepare_format_string(format: &str, date_format_style: &str) -> String {
...
+                ts_values.push(0);
+                offset_values.push(0);
+            }
+        }
+    }
+
+    (ts_values, offset_values, len)
+}
+
+fn timestamp_from_parts_fn(
```
Comment (TCeason) [inline src/query/functions/src/scalars/timestamp/src/datetime.rs:2972]: Why not call clamp_timestamp check timestamp?

File: src/query/expression/src/utils/date_helper.rs [hunk 21]
```diff
@@ -2066,3 +2067,109 @@ pub fn pg_format_to_strftime(pg_format_string: &str) -> String {
...
+    let norm_month = (total_months.rem_euclid(12) + 1) as i8; // 1..=12
+
+    let norm_year =
+        i16::try_from(norm_year_i64).map_err(|_| format!("Year out of bounds: {norm_year_i64}"))?;
+
+    // Create base date
+    let base = Date::new(norm_year, norm_month, 1)
+        .map_err(|_| format!("Invalid date: year={year}, month={month}, day={day}"))?;
+
+    // 2. Safe day normalization (Fixing the second bot flag)
```
Comment (TCeason) [inline src/query/expression/src/utils/date_helper.rs:2091]: need to delete AI comment.

File: tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test [hunk 22]
```diff
@@ -1812,3 +1812,143 @@ query ??
...
+
+# Nanosecond overflow with timezone
+query ?
+select timestamp_tz_from_parts(2023, 1, 1, 12, 0, 0, 2000000000, 'UTC');
+----
+2023-01-01 12:00:02.000000 +0000
+
+# Negative nanoseconds
+query ?
+select timestamp_tz_from_parts(2023, 1, 1, 12, 0, 0, -1000000000, 'UTC');
```
Comment (TCeason) [inline tests/sqllogictests/suites/query/functions/02_0012_function_datetimes.test:1952]: Please add test like this:

```sql
  -- date_from_parts with table data and NULLs                                                                                    
  statement ok                                                    
  CREATE TABLE t_date_parts(year INT64 NULL, month INT64 NULL, day INT64 NULL)

  statement ok
  INSERT INTO t_date_parts VALUES (2023, 6, 15), (2010, 1, 100), (NULL, 6, 1), (2023, NULL, 1), (2023, 6, NULL)

  query ?
  SELECT date_from_parts(year, month, day) FROM t_date_parts ORDER BY year, month, day
  ----
  NULL
  NULL
  2010-04-10
  2023-06-15

  -- timestamp_from_parts with table data and NULLs
  statement ok
  CREATE TABLE t_ts_parts(year INT64 NULL, month INT64 NULL, day INT64 NULL, hour INT64 NULL, minute INT64 NULL, second INT64
  NULL)

  statement ok
  INSERT INTO t_ts_parts VALUES (2013, 4, 5, 12, 0, 0), (NULL, 4, 5, 12, 0, 0), (2013, 4, 5, NULL, 0, 0)

  query ?
  SELECT timestamp_from_parts(year, month, day, hour, minute, second) FROM t_ts_parts ORDER BY year, month
  ----
  NULL
  NULL
  2013-04-05 12:00:00.000000

  -- timestamp_tz_from_parts with table data and NULLs
  statement ok
  CREATE TABLE t_tstz_parts(year INT64 NULL, month INT64 NULL, day INT64 NULL, hour INT64 NULL, minute INT64 NULL, second INT64
  NULL, tz VARCHAR NULL)

  statement ok
  INSERT INTO t_tstz_parts VALUES (2023, 1, 1, 12, 0, 0, 'UTC'), (2023, 7, 1, 12, 0, 0, 'Asia/Shanghai'), (NULL, 1, 1, 12, 0, 0,
  'UTC'), (2023, 1, 1, 12, 0, 0, NULL)

  query ?
  SELECT timestamp_tz_from_parts(year, month, day, hour, minute, second, tz) FROM t_tstz_parts ORDER BY year, tz
  ----
  NULL
  NULL
  2023-01-01 12:00:00.000000 +0000
  2023-07-01 12:00:00.000000 +0800

  statement ok
  DROP TABLE t_date_parts

  statement ok
  DROP TABLE t_ts_parts

  statement ok
  DROP TABLE t_tstz_parts
```
